### PR TITLE
Make imageio an optional dependency

### PR DESCRIPTION
**Pull request by basnijholt**
### Overview

As discussed in https://github.com/pyvista/pyvista/issues/4302 and https://github.com/conda-forge/pyvista-feedstock/pull/76.

### Bonus

- Discovered that `imageio==2.28.0` doesn't write multiple frames, so let's keep it `<2.28.0` for now. Reported in https://github.com/imageio/imageio/issues/974

### CODE REVIEW

**Review by banesullivan — APPROVED** [review]
LGTM!

**Review by tkoyama010 — APPROVED** [review]

File: pyvista/plotting/plotting.py [hunk 1]
```diff
@@ -4842,7 +4844,12 @@ def open_movie(self, filename, framerate=24, quality=5, **kwargs):
         >>> pl.open_movie('movie.mp4', quality=10)  # doctest:+SKIP
 
         """
-        from imageio import get_writer
+        try:
+            from imageio import get_writer
+        except ModuleNotFoundError:  # pragma: no cover
+            raise ModuleNotFoundError(
+                'Install imageio to use `open_movie` with:\n\n' '   pip install imageio'
```
Comment (akaszynski) [inline pyvista/plotting/plotting.py:4851]:
```suggestion
                'Install imageio to use `open_movie` with:\n\n   pip install imageio'
```

File: pyvista/plotting/plotting.py [hunk 2]
```diff
@@ -4899,7 +4908,12 @@ def open_gif(self, filename, loop=0, fps=10, palettesize=256, subrectangles=Fals
         See :ref:`gif_movie_example` for a full example using this method.
 
         """
-        from imageio import get_writer
+        try:
+            from imageio import get_writer
+        except ModuleNotFoundError:  # pragma: no cover
+            raise ModuleNotFoundError(
+                'Install imageio to use `open_gif` with:\n\n' '   pip install imageio'
```
Comment (akaszynski) [inline pyvista/plotting/plotting.py:4915]:
```suggestion
                'Install imageio to use `open_gif` with:\n\n   pip install imageio'
```

File: pyproject.toml [hunk 3]
```diff
@@ -38,18 +37,30 @@ dynamic = ['version']
 
 [project.optional-dependencies]
 all = [
+    'imageio',
     'cmocean',
     'colorcet',
+    'imageio',
```
Comment (basnijholt) [inline pyproject.toml:43]: `imageio` is duplicated

File: pyvista/utilities/misc.py [hunk 4]
```diff
@@ -35,6 +35,38 @@ def _check_range(value, rng, parm_name):
...
+    Raises
+    ------
+    ModuleNotFoundError
+        Raised when ``imageio`` is not installed when attempting to read
+        ``filename``.
+
+    Returns
+    -------
+    imageio.core.util.Array
+        Image read from ``imageio``.
```
Comment (akaszynski) [inline pyvista/utilities/misc.py:55]:
```suggestion
    Returns
    -------
    imageio.core.util.Array
        Image read from ``imageio``.

    Raises
    ------
    ModuleNotFoundError
        Raised when ``imageio`` is not installed when attempting to read
        ``filename``.
```

File: pyproject.toml (hunk 3 above)
Comment (akaszynski) [reply pyproject.toml:43]:
```suggestion
    'imageio',
    'cmocean',
    'colorcet',
```

**Review by banesullivan — APPROVED** [review]

File: pyvista/utilities/misc.py [hunk 5]
```diff
@@ -35,6 +35,38 @@ def _check_range(value, rng, parm_name):
         )
 
 
+def try_imageio_imread(filename):
```
Comment (MatthewFlamm) [inline pyvista/utilities/misc.py:38]: Will this function end up in the pyvista namesapce?  Do we want it to?  It is mostly an internal PyVista function that checks for importability, so it seems not that useful for downstream users.  If this currently ending up in the namespace, then I'd argue that we make this private instead.

Comment (akaszynski) [reply pyvista/utilities/misc.py:38]: Shoot. I'll fix this.

**Review by akaszynski — DISMISSED** [review]

**Comment by basnijholt** [comment]
@akaszynski, thanks a lot for doing most of the work! 😄

**Review by tkoyama010 — DISMISSED** [review]

**Review by akaszynski — APPROVED** [review]

**Pull request merged**